addons/website/models/website_visitor.py

```diff
@@ -0,0 +1,129 @@
+    lang_id = fields.Many2one('res.lang', string='Language', help="Language from the website when visitor has been created")
+    visit_count = fields.Integer('Number of visits', default=1, readonly=True, help="A new visit is considered if last connection was more than 8 hours ago.")
+    page_ids = fields.One2many('website.visitor.page', 'visitor_id', string='Visited Pages History')
+    page_view_count = fields.Integer('Page Views', compute="_compute_page_view_count")
+    unique_page_ids = fields.Many2many('website.page', string="Visited Pages", compute="_compute_unique_page_ids", store=True)
+    time_last_action = fields.Float('Time since last action', compute="_compute_time_last_action", help='Time since last page view in seconds')
+    is_connected = fields.Boolean('Is connected ?', compute='_compute_is_connected', help='A visitor is considered as connected if his last page view was within the last 5 minutes.')
+
+    @api.depends('partner_ids')
+    def _compute_partner_count(self):
+        for visitor in self:
+            visitor.partner_count = len(visitor.partner_ids)
+
+    @api.depends('page_ids')
+    def _compute_page_view_count(self):
+        for visitor in self:
+            visitor.page_view_count = len(visitor.page_ids)
+
+    @api.depends('page_ids')
+    def _compute_unique_page_ids(self):
+        for visitor in self:
+            unique_page_ids = list(set(visitor.page_ids.page_id.ids))
+            visitor.unique_page_ids = [(6, 0, unique_page_ids)]
+
+    @api.depends('last_connection_date')
+    def _compute_time_last_action(self):
+        for visitor in self:
+            visitor.time_last_action = (datetime.now() - visitor.last_connection_date).total_seconds()/60
+
+    @api.depends('last_connection_date')
+    def _compute_is_connected(self):
+        for visitor in self:
+            visitor.is_connected = (datetime.now() - visitor.last_connection_date) < timedelta(minutes=5)
+
+    def _encode(self, visitor_id):
+        md5_visitor_id = md5(b"%d%s" % (visitor_id, self._get_key().encode('ascii'))).hexdigest()
+        return "%d-%s" % (visitor_id, md5_visitor_id)
+
+    def _decode(self):
+        """ Opens the cookie, verifies the signature of the visitor
+        :return: visitor_id if the verification passes, None otherwise"""
+        if request:  # To pass the tests (as in tests, users are authenticated BEFORE making requests)
+            cookie_content = request.httprequest.cookies.get('visitor_id')
+            if cookie_content and '-' in cookie_content:
+                visitor_id, md5_visitor_id = cookie_content.split('-', 1)
+                expected_encryped_visitor_id = md5(("%s%s" % (visitor_id, self._get_key())).encode('utf-8')).hexdigest()
+                if md5_visitor_id == expected_encryped_visitor_id:
+                    visitor = int(visitor_id)
+                    return visitor if visitor else None
+        return None
+
+    def _get_key(self):
```

Comment:
David !

I think it is better to have a _compute_hash on visitor model (taking visitor_id) parameter and doing the hash. That way you hide complexity of hash, not getting the database secret ^^ .